tests/test_edge_cases.py

```diff
@@ -2634,3 +2635,71 @@ class E(BaseModel):
 
     # This is just to check that above model contraption doesn't fail
     assert E(c=C(b=B(a_1=A(), a_2=A()))).model_dump() == {'c': {'b': {'a_1': {}, 'a_2': {}}}}
+
+
+def test_recursive_root_models_in_discriminated_union():
+    class Model1(BaseModel):
+        kind: Literal['1'] = '1'
+        two: 'Model2 | None'
+
+    class Model2(BaseModel):
+        kind: Literal['2'] = '2'
+        one: Model1 | None
+
+    class Root1(RootModel[Model1]):
+        pass
+
+    class Root2(RootModel[Model2]):
+        pass
+
+    class Outer(BaseModel):
+        first: Annotated[Root1 | Root2, Field(discriminator='kind')]
+        second: Annotated[Root1 | Root2, Field(discriminator='kind')]
+
+    assert Outer.model_json_schema() == {
```

Comment:
Prior to #7646, this test fails to even create the `Outer` class due to a `SchemaError`. I added the `model_json_schema()` call though to make sure the generated JSON schema is also correct (which it appears to be), as an additional layer of protection against bugs that might get introduced in the future.